cpp/tests/filter/filter_test.cpp

```diff
@@ -272,6 +273,105 @@ __device__ void filter(bool* out,
+      .release();
+  std::unique_ptr<cudf::column> bool_a =
+    cudf::test::fixed_width_column_wrapper<bool>{
+      {false, false, true, true, false, false, true, true}, {1, 1, 1, 1, 1, 0, 0, 0}}
+      .release();
+  std::unique_ptr<cudf::column> bool_b =
+    cudf::test::fixed_width_column_wrapper<bool>{
+      {false, true, false, true, true, true, false, true}, {1, 1, 1, 1, 0, 1, 1, 0}}
+      .release();
+
+  cudf::table_view table = cudf::table_view({a->view(), b->view(), bool_a->view(), bool_b->view()});
+};
+
+using Executors = cudf::test::Types<ast_expression_executor, jit_expression_executor>;
+
+TYPED_TEST_SUITE(FilterExpressionTest, Executors);
+
+TYPED_TEST(FilterExpressionTest, IsNull)
+{
+  using Executor = TypeParam;
+
+  auto tree          = cudf::ast::tree();
+  auto& ref_0        = tree.push(cudf::ast::column_reference(0));
+  auto& is_null_expr = tree.push(cudf::ast::operation(cudf::ast::ast_operator::IS_NULL, ref_0));
+  auto& filter_expr  = tree.push(cudf::ast::operation(cudf::ast::ast_operator::NOT, is_null_expr));
+  auto result        = Executor::filter(filter_expr, this->table);
+  auto expected      = cudf::test::fixed_width_column_wrapper<int32_t>{1, 2, 3, 4, 5, 6, 7};
+  CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT(expected, result->get_column(0).view());
+}
+
+TYPED_TEST(FilterExpressionTest, NullEqual)
+{
+  using Executor = TypeParam;
+
+  auto tree   = cudf::ast::tree();
+  auto& ref_0 = tree.push(cudf::ast::column_reference(0));
+  auto& ref_1 = tree.push(cudf::ast::column_reference(1));
+  auto& null_equal_expr =
+    tree.push(cudf::ast::operation(cudf::ast::ast_operator::NULL_EQUAL, ref_0, ref_1));
+  auto result = Executor::filter(null_equal_expr, this->table);
+  auto expected =
+    cudf::test::fixed_width_column_wrapper<int32_t>{{1, 3, 4, 5, 6, 7, 8}, {1, 1, 1, 1, 1, 1, 0}};
+  CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT(expected, result->get_column(0).view());
+}
+
+TYPED_TEST(FilterExpressionTest, NullLogicalAnd)
+{
+  using Executor = TypeParam;
+
+  auto tree   = cudf::ast::tree();
+  auto& ref_2 = tree.push(cudf::ast::column_reference(2));
+  auto& ref_3 = tree.push(cudf::ast::column_reference(3));
+  auto& and_expr =
+    tree.push(cudf::ast::operation(cudf::ast::ast_operator::NULL_LOGICAL_AND, ref_2, ref_3));
+  auto result   = Executor::filter(and_expr, this->table);
+  auto expected = cudf::test::fixed_width_column_wrapper<int32_t>{4};
+  CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT(expected, result->get_column(0).view());
+}
+
+TYPED_TEST(FilterExpressionTest, NullLogicalOr)
+{
+  using Executor = TypeParam;
+
+  auto tree   = cudf::ast::tree();
+  auto& ref_2 = tree.push(cudf::ast::column_reference(2));
+  auto& ref_3 = tree.push(cudf::ast::column_reference(3));
+  auto& or_expr =
+    tree.push(cudf::ast::operation(cudf::ast::ast_operator::NULL_LOGICAL_OR, ref_2, ref_3));
+  auto result   = Executor::filter(or_expr, this->table);
+  auto expected = cudf::test::fixed_width_column_wrapper<int32_t>{2, 3, 4, 6};
```

Comment:
I didn't understand this one. Shouldn't we get a nullable boolean column output here. Same with `NullLogicalAnd`?

Reply:
It's an equivalence test, not strict equality.
I don't think it matters if it produces a nullable column or not. As long as the values are equivalent.
It is still consistent with our null-mask invariants.